stdlib/public/Concurrency/Actor.cpp

```diff
@@ -1498,56 +1389,89 @@ void DefaultActorImpl::enqueueStealer(Job *job, JobPriority priority) {
+  }
 
+  // We must ensure that any jobs not seen by the collectJobs() don't have any
+  // dangling references to the jobs that have been collected. For that we must
+  // atomically set head pointer to NULL. If it fails because more jobs have
+  // been added in the meantime, we have to continue collecting jobs.
   while (true) {
     assert(oldState.isAnyRunning());
 
-    if (!firstJob) {
-      // Nothing to drain, short circuit
-      SWIFT_TASK_DEBUG_LOG("No jobs to drain on actor %p", this);
-      return NULL;
-    }
-
     auto newState = oldState;
-    // Dequeue the first job and set up a new head
-    newState = newState.withFirstJob(getNextJobInQueue(firstJob));
-    if (_status().compare_exchange_weak(oldState, newState,
-                            /* success */ std::memory_order_relaxed,
-                            /* failure */ std::memory_order_relaxed)) {
-      SWIFT_TASK_DEBUG_LOG("Drained first job %p from actor %p", firstJob, this);
-      traceActorStateTransition(this, oldState, newState, distributedActorIsRemote);
-      concurrency::trace::actor_dequeue(this, firstJob);
-      return firstJob;
+    newState = newState.withFirstUnprioritisedJob(nullptr);
+
+    if (_status().compare_exchange_weak(
+            oldState, newState,
+            /* success */ std::memory_order_relaxed,
+            /* failure */ std::memory_order_relaxed)) {
+      SWIFT_TASK_DEBUG_LOG("Collected some jobs from actor %p", this);
+      traceActorStateTransition(this, oldState, newState,
+                                distributedActorIsRemote);
+      return result;
     }
 
     // We failed the weak cmpxchg spuriously, go through loop again.
-    if (oldState.getFirstJob().getAsJob() == jobToPreprocessFrom.getAsJob()) {
+    if (oldState.getFirstUnprioritisedJob() == result.tail) {
       continue;
     }
 
     // There were new items concurrently added to the queue. We need to
-    // preprocess the newly added unprocessed items and merge them to the already
-    // preprocessed list.
+    // collect the newly added unprocessed items and append them to the
+    // already collected list.
     //
-    // The newly merged items that need to be preprocessed, are between the head
-    // of the linked list, and the last job we did the previous preprocessQueue
-    // on
-    firstJob = preprocessQueue(oldState.getFirstJob(), jobToPreprocessFrom, firstJob);
-    jobToPreprocessFrom = oldState.getFirstJob();
-    traceJobQueue(this, firstJob);
+    job = oldState.getFirstUnprioritisedJob();
+    SimpleQueue extraJobs;
+    while (job != result.tail) {
+      auto next = getNextJob(job);
+      extraJobs.prepend(job);
+      job = next;
+    }
+    result.appendContentsOf(extraJobs);
+  }
+}
+
+// Called with actor lock held on current thread
+void DefaultActorImpl::processJobs() {
+  SimpleQueue jobs = collectJobs();
+  auto job = jobs.head;
+  while (job) {
+    auto next = getNextJob(job);
+    prioritisedJobs.enqueue(job);
```

Comment:
This operation would be faster if we just gave prioritizedJobs the whole queue at once — it's common to get runs of jobs with the same priority, and we don't actually need to rewrite the internal links of those runs.